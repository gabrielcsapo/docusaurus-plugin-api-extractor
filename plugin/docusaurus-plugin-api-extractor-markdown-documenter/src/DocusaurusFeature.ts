import {
  MarkdownDocumenterFeature,
  IMarkdownDocumenterFeatureOnBeforeWritePageArgs,
  MarkdownDocumenterAccessor
} from '@microsoft/api-documenter';
import { ApiItem, ApiItemKind } from '@microsoft/api-extractor-model';
import fs from 'fs';
import ejs from 'ejs';
import { writeFileSync } from 'fs';
import prettier from 'prettier';
import { parse, join } from 'path';
import { ICategoryNode, IDocNode } from './interfaces';
import { toDocusaurusMarkDown } from './markdown/to-docusaurus-markdown';

const sidebar: string = fs.readFileSync(join(__dirname, './api-sidebar.ejs'), 'utf-8');
const tree: string = fs.readFileSync(join(__dirname, './api-sidebar-tree.ejs'), 'utf-8');
// eslint-disable-next-line @typescript-eslint/typedef
const sidebarTmpl = ejs.compile(sidebar);
// eslint-disable-next-line @typescript-eslint/typedef
const treeTmpl = ejs.compile(tree);

export class DocusaurusFeature extends MarkdownDocumenterFeature {
  public onBeforeWritePage(eventArgs: IMarkdownDocumenterFeatureOnBeforeWritePageArgs): void {
    const { name: id } = parse(eventArgs.outputFilename);

    eventArgs.pageContent = toDocusaurusMarkDown(eventArgs.pageContent, id);
  }

  public onFinished(): void {
    const navigationFile: ICategoryNode[] = [];
    const packages: ICategoryNode = {
      type: 'category',
      label: 'Packages',
      items: [overviewNode('index')],
      collapsed: false
    };
    navigationFile.push(packages);

    buildNavigation(packages.items!, this.context.apiModel, this.context.documenter);

    const sidebarFile: string = prettier.format(
      sidebarTmpl({
        sideBarItems: navigationFile,
        dir: 'api',
        isSideBarItem,
        tree: treeTmpl
      }),
      { parser: 'babel', singleQuote: true }
    );

    writeFileSync(join(this.context.outputFolder, 'api-sidebar.js'), sidebarFile);
  }
}

function isSideBarItem(items: string[] | ICategoryNode[]): boolean {
  return Array.isArray(items) && typeof items[0] === 'object' && items[0] !== null && 'type' in items[0];
}

export function buildNavigation(
  parentNodes: unknown[],
  parentAPIItem: ApiItem,
  documenter: MarkdownDocumenterAccessor
): void {
  if (!parentAPIItem.members) return;
  for (const apiItem of parentAPIItem.members) {
    const kind: ApiItemKind = apiItem.kind;

    const id: string = `${documenter.getLinkForApiItem(apiItem)?.replace('./', '')?.replace('.md', '')}`;

    let navItem: ICategoryNode;
    switch (kind) {
      case 'EntryPoint':
        buildNavigation(parentNodes, apiItem, documenter);
        break;
      case 'Package':
      case 'Class':
      case 'Interface':
      case 'Namespace':
        navItem = {
          type: 'category',
          label: apiItem.displayName,
          collapsed: shouldCollapse(kind),
          items: [overviewNode(id)]
        };
        parentNodes.push(navItem);
        buildNavigation(navItem.items, apiItem, documenter);
        break;
      default:
        parentNodes.push({
          type: 'doc',
          label: apiItem.displayName,
          id
        });
        buildNavigation(parentNodes, apiItem, documenter);
    }
  }
}

function shouldCollapse(kind: ApiItemKind): boolean {
  return kind === 'Class' || kind === 'Namespace' || kind === 'Interface';
}

function overviewNode(id: string): IDocNode {
  return {
    type: 'doc',
    label: 'Overview',
    id
  };
}
