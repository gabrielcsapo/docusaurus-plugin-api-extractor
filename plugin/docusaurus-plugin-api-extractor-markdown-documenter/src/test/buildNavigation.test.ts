import { buildNavigation } from '../DocusaurusFeature';
import { join } from 'path';
import { ApiItem, ApiModel } from '@microsoft/api-extractor-model';
import type { MarkdownDocumenterAccessor } from '@microsoft/api-documenter';

const documenter = {
  getLinkForApiItem(apiModel: ApiItem) {
    return apiModel.displayName;
  }
} as MarkdownDocumenterAccessor;

it('it produces a tree structure for the sidebar', () => {
  const model = new ApiModel();
  model.loadPackage(join(__dirname, './fixtures/api-model.json'));
  const parentNodes: unknown[] = [];
  buildNavigation(parentNodes, model, documenter);
  expect(parentNodes).toMatchSnapshot();
});
