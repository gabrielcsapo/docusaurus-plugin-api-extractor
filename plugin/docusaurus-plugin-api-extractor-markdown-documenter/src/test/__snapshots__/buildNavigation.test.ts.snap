// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it produces a tree structure for the sidebar 1`] = `
Array [
  Object {
    "collapsed": false,
    "items": Array [
      Object {
        "id": "example-project",
        "label": "Overview",
        "type": "doc",
      },
      Object {
        "collapsed": true,
        "items": Array [
          Object {
            "id": "Animal",
            "label": "Overview",
            "type": "doc",
          },
          Object {
            "id": "speak",
            "label": "speak",
            "type": "doc",
          },
        ],
        "label": "Animal",
        "type": "category",
      },
      Object {
        "collapsed": true,
        "items": Array [
          Object {
            "id": "AnimalOptions",
            "label": "Overview",
            "type": "doc",
          },
          Object {
            "id": "name",
            "label": "name",
            "type": "doc",
          },
        ],
        "label": "AnimalOptions",
        "type": "category",
      },
      Object {
        "id": "bar",
        "label": "bar",
        "type": "doc",
      },
      Object {
        "collapsed": true,
        "items": Array [
          Object {
            "id": "Bizz",
            "label": "Overview",
            "type": "doc",
          },
          Object {
            "id": "(constructor)",
            "label": "(constructor)",
            "type": "doc",
          },
          Object {
            "id": "bar",
            "label": "bar",
            "type": "doc",
          },
          Object {
            "id": "foo",
            "label": "foo",
            "type": "doc",
          },
        ],
        "label": "Bizz",
        "type": "category",
      },
      Object {
        "collapsed": true,
        "items": Array [
          Object {
            "id": "Box",
            "label": "Overview",
            "type": "doc",
          },
          Object {
            "id": "(constructor)",
            "label": "(constructor)",
            "type": "doc",
          },
          Object {
            "id": "unwrap",
            "label": "unwrap",
            "type": "doc",
          },
        ],
        "label": "Box",
        "type": "category",
      },
      Object {
        "id": "foo",
        "label": "foo",
        "type": "doc",
      },
      Object {
        "id": "sum",
        "label": "sum",
        "type": "doc",
      },
    ],
    "label": "example-project",
    "type": "category",
  },
]
`;
